3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,202 @@
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventsListeners_.push(listener);
+  }
+
+  /**
+   * Registers a callback function to be called when extra setup data is
+   * recevied from an AMP Analytics tag.
+   * @param {!Function} listener A function to receive any extra
+   * setup data sent by the creative
+   */
+  registerAmpAnalytics3pExtraDataListener(listener) {
+    this.extraDataListeners_.push(listener);
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @param {!string} message The message to send.
+   */
+  sendMessageToCreative(message) {
+    this.iframeMessagingClient_.sendMessage('ampAnalytics3pResponse',
+      {senderId: this.id_, ampAnalytics3pResponse: message});
+  }
+
+  /**
+   * Gets any optional extra data that should be made available to the
+   * cross-domain frame, in the context of a particular creative.
+   * @param {number} senderId The ID of the creative that sent the extra data
+   * @returns {?string}
+   */
+  getExtraData(senderId) {
+    return this.senderIdToExtraData_[senderId];
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage('ampAnalytics3pReady',
+      {senderId: this.id_});
+  }
+
+  /**
+   * Handles messages received from the creative, by enqueueing them to
+   * be sent to the third-party vendor's metrics-collection page
+   * @param {!Object} received The messages that were received from the creative
+   * @private
+   */
+  onReceivedMessagesFromCreative_(received) {
+    if (!received || !received.ampAnalytics3pMessages) {
+      return;
+    }
+    received.ampAnalytics3pMessages.forEach(submessage => {
+      if (submessage.ampAnalytics3pEvent) {
+        this.receivedMessagesQueue_.push(submessage);
+      } else if (submessage.ampAnalytics3pExtraData) {
+        this.senderIdToExtraData_[submessage.senderId] =
+          submessage.ampAnalytics3pExtraData;
+        this.receivedExtraDataQueue_.push(submessage);
+      }
+    });
+    this.sendQueuedMessagesToListeners_();
+  }
+
+  /**
+   * Sends enqueued messages to the third-party vendor's metrics-collection page
+   * If there are no event listeners on the page, event messages will remain
+   * in the queue. But if there is at least one event listener, messages will be
+   * sent to those listener(s) and the queue will be emptied. This means
+   * that the first event listener will receive messages that predate its
+   * creation, but if there are additional event listeners, they may not.
+   * Likewise for extra data listeners.
+   * @private
+   */
+  sendQueuedMessagesToListeners_() {
+    this.flushQueue_(this.receivedMessagesQueue_, this.eventsListeners_);
```

Comment:
Given flush queue is an instance function, its better to let it reset the queue instead of doing so here (as it may want to leave the queue in place for certain conditions in the future).